### PR TITLE
feat(grafast/grafserv/src/servers/deno.ts): add deno server

### PR DESCRIPTION
**Pull request by nestarz**
A first stop to be used in Deno.

### CODE REVIEW

**Comment by benjie** [comment]
(Just a note to say thank you for submitting this. I'm currently super busy preparing for GraphQLConf so I might not get around to properly evaluating this until October - not least because I've not really used Deno before and will have to familiarize myself with it before taking on the maintenance of this adaptor.)

File: grafast/grafserv/src/servers/deno.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      >;
+    },
+    async getBody() {
+      return {
+        type: "json",
+        json: await req.json(),
+      } as GrafservBodyJSON;
+    },
+    requestContext: { ctx },
+  };
```
Comment (benjie) [inline grafast/grafserv/src/servers/deno.ts:43]: We should avoid calling `new URL()` many times when we could call it just once. I've also added a couple TODOs but they don't need to be addressed in this PR.

```suggestion
  const url = new URL(req.url);
  return {
    // TODO: figure out the actual HTTP version
    httpVersionMajor: 2,
    httpVersionMinor: 0,
    isSecure: url.protocol === "https",
    method: req.method,
    path: url.pathname,
    headers: Object.fromEntries(req.headers.entries()),
    getQueryParams() {
      // TODO: on duplicate keys, this should convert to an array instead
      return Object.fromEntries(url.searchParams) as Record<
        string,
        string | string[]
      >;
    },
    async getBody() {
      return {
        type: "json",
        json: await req.json(),
      } as GrafservBodyJSON;
    },
    requestContext: {
      denov1: {
        ctx,
      },
    },
  };
```

File: grafast/grafserv/src/servers/deno.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  GrafservBodyJSON,
+  GrafservConfig,
+  RequestDigest,
+  Result,
+} from "grafserv";
+
+declare global {
+  namespace Grafast {
+    interface RequestContext {
+      ctx: Deno.ServeHandlerInfo;
```
Comment (benjie) [inline grafast/grafserv/src/servers/deno.ts:17]: This needs to be namespaced to Deno:

```suggestion
      denov1: {
        ctx: Deno.ServeHandlerInfo;
      };
```

**Review by benjie — CHANGES_REQUESTED** [review]
This looks great; I love how simple it is!

File: grafast/grafserv/src/servers/deno/v1/index.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      },
+    },
+  };
+}
+
+export class DenoGrafserv extends GrafservBase {
+  constructor(config: GrafservConfig) {
+    super(config);
+  }
+  public async handler(req: Request, ctx: Deno.ServeHandlerInfo) {
```
Comment (benjie) [inline grafast/grafserv/src/servers/deno/v1/index.ts:59]: The constructor wasn't needed.

```suggestion
  public async handler(req: Request, ctx: Deno.ServeHandlerInfo) {
```

**Review by benjie — CHANGES_REQUESTED** [review]
That's as much as I can do right now; please:

1. Solve the issue so that TypeScript running on Node doesn't crash (install the Deno types?)
2. Add a page to the Grafast documentation like this one: https://github.com/benjie/crystal/blob/main/grafast/website/grafserv/servers/fastify.md (Ideally better!)

**Comment by nestarz** [comment]
Thanks for your review ! I will add the required changes you mention asap. Thanks for your work.

**Comment by benjie** [comment]
(Please be aware this PR should now be raised against https://github.com/graphile/crystal instead :heart:)

**Comment by benjie** [comment]
Please re-open this against the above repo. I've pushed a few more commits but I've literally no idea how to go about running this locally - it seems that Deno doesn't support running local code in `npm` compatibility mode? https://github.com/denoland/deno/issues/18474 Without that, I can't be certain it works, so that's the first thing to solve.

**Comment by nestarz** [comment]
Noted ! Thanks for your work @benjie

**Comment by benjie** [comment]
Incidentally if you need help doing the git-fu to move this over:

1. Pull down these commits into your local clone (`git pull --rebase`)
2. Fork the `graphile/crystal` repo via GitHub interface
3. Change the remote for your existing local clone to point at the new repository (`git remote set-url origin git@github.com:nestarz/<new repo name here>.git`)
4. `git push -u origin patch-1`

Then raise a PR :)

**Pull request closed without merging**